safer-ffi-gen-macro/src/enum_to_error_code.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{Error, ErrorReason};
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::{Attribute, Ident, ItemEnum};
+
+pub fn impl_enum_to_error_code(enumeration: ItemEnum) -> Result<TokenStream, Error> {
+    let cfgs = enumeration
+        .attrs
+        .iter()
+        .filter(|&attr| is_cfg(attr))
+        .collect::<Vec<_>>();
+
+    let ty = &enumeration.ident;
+    let (impl_generics, ty_generics, where_clause) = enumeration.generics.split_for_impl();
+    let discriminant_type = Ident::new(&format!("{ty}Code"), Span::call_site());
+
+    let non_exhaustive = if enumeration.attrs.iter().any(is_non_exhaustive) {
+        Some(quote! { #[non_exhaustive] })
+    } else {
+        None
+    };
+
+    let discriminant_variants = enumeration
+        .variants
+        .iter()
+        .enumerate()
+        .map(|(i, v)| {
+            let i =
+                index_to_error_code(i).ok_or_else(|| ErrorReason::TooManyVariants.spanned(ty))?;
+            let cfgs = v.attrs.iter().filter(|&attr| is_cfg(attr));
+            let variant = &v.ident;
+
+            Ok(quote! {
+                #(#cfgs)*
+                #variant = #i
+            })
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let branches = enumeration.variants.iter().map(|v| {
+        let cfgs = v.attrs.iter().filter(|&attr| is_cfg(attr));
+        let variant = &v.ident;
+
+        quote! {
+            #(#cfgs)*
+            #ty::#variant { .. } => #discriminant_type::#variant as i32
+        }
+    });
+
+    Ok(quote! {
+        #(#cfgs)*
+        impl #impl_generics ::core::convert::From<&#ty #ty_generics> for i32 #where_clause {
+            fn from(x: &#ty #ty_generics) -> i32 {
+                match *x {
+                    #(#branches,)*
+                }
+            }
+        }
+
+        #(#cfgs)*
+        #[::safer_ffi_gen::safer_ffi::derive_ReprC]
+        #[repr(i32)]
+        #non_exhaustive
+        pub enum #discriminant_type {
+            #(#discriminant_variants,)*
+        }
+    })
+}
+
+fn is_cfg(attr: &Attribute) -> bool {
+    attr_is(attr, "cfg")
+}
+
+fn is_non_exhaustive(attr: &Attribute) -> bool {
+    attr_is(attr, "non_exhaustive")
+}
+
+fn attr_is(attr: &Attribute, id: &str) -> bool {
+    attr.path().get_ident().map_or(false, |i| i == id)
+}
+
+fn index_to_error_code(i: usize) -> Option<i32> {
```

Comment:
Does it make any difference in final code size for i32 vs u8? I don't think it would but just checking